vtex/faststore
# feat(store-api): Add cart to store-api

## What's the purpose of this pull request?
This PR adds an optimistic cart implementation to store-api.

## How it works? 
The optimistic cart we have works via state validation. This means that, we build a state of the cart on the browser, and then, the backend validates this state. If this state is valid, the backend signalizes this state is ok and nothing changes on the frontend. If, however, this state is not valid, a.k.a, this product is unavailable, a promotion was applied, etc, the backend sends the new cart state to the frontend. An example of this process can be seen below: 

![image](https://user-images.githubusercontent.com/1753396/134071242-86b2e100-408b-4b85-81f7-85438d4774e5.png)

To implement the cart, I used the schema.org's type `Order`. Also, another thing this PR brings is `input` fields. We didn't really use them before, but now I needed a standard for writing them. I added an `I` prefix to them, so writing an `input` for a `type` can be achieved by:
`type StoreProduct` -> `type IStoreProduct`.

Also, another thing this PR brings is the usage of [dataloaders](https://github.com/graphql/dataloader). This was necessary for loading skus and simulations without overwhelming the search/checkout API.

## How to test it?
To test it, open our PR on base.store and check that all promotions are applied correctly. Also, if you want you can change the base account name to `carrefourbrfood` and check all edge cases in there. 

https://github.com/vtex-sites/base.store/pull/36
